src/request/plan.rs

```diff
@@ -0,0 +1,303 @@
+        input.into_iter().collect()
+    }
+}
+
+/// Process data into another kind of data.
+pub trait Process: Sized + Clone + Send + Sync + 'static {
+    type Out: Send;
+
+    fn process(input: Result<Self>) -> Result<Self::Out>;
+}
+
+#[derive(Clone)]
+pub struct ProcessResponse<P: Plan, Pr: Process> {
+    pub inner: P,
+    pub phantom: PhantomData<Pr>,
+}
+
+#[async_trait]
+impl<P: Plan<Result = Pr>, Pr: Process> Plan for ProcessResponse<P, Pr> {
+    type Result = Pr::Out;
+
+    async fn execute(&self) -> Result<Self::Result> {
+        Pr::process(self.inner.execute().await)
+    }
+}
+
+pub struct RetryRegion<P: Plan, PdC: PdClient> {
+    pub inner: P,
+    pub pd_client: Arc<PdC>,
+    pub backoff: Backoff,
+}
+
+impl<P: Plan, PdC: PdClient> Clone for RetryRegion<P, PdC> {
+    fn clone(&self) -> Self {
+        RetryRegion {
+            inner: self.inner.clone(),
+            pd_client: self.pd_client.clone(),
+            backoff: self.backoff.clone(),
+        }
+    }
+}
+
+#[async_trait]
+impl<P: Plan, PdC: PdClient> Plan for RetryRegion<P, PdC>
+where
+    P::Result: HasError,
+{
+    type Result = P::Result;
+
+    async fn execute(&self) -> Result<Self::Result> {
+        let mut result = self.inner.execute().await?;
+        let mut clone = self.clone();
+        while let Some(region_error) = result.region_error() {
+            match clone.backoff.next_delay_duration() {
+                None => return Err(region_error),
+                Some(delay_duration) => {
+                    futures_timer::Delay::new(delay_duration).await;
+                    result = clone.inner.execute().await?;
+                }
+            }
+        }
+
+        Ok(result)
+    }
+}
+
+pub struct ResolveLock<P: Plan, PdC: PdClient> {
+    pub inner: P,
+    pub pd_client: Arc<PdC>,
+    pub backoff: Backoff,
+}
+
+impl<P: Plan, PdC: PdClient> Clone for ResolveLock<P, PdC> {
+    fn clone(&self) -> Self {
+        ResolveLock {
+            inner: self.inner.clone(),
+            pd_client: self.pd_client.clone(),
+            backoff: self.backoff.clone(),
+        }
+    }
+}
+
+#[async_trait]
+impl<P: Plan, PdC: PdClient> Plan for ResolveLock<P, PdC>
+where
+    P::Result: HasLocks,
+{
+    type Result = P::Result;
+
+    async fn execute(&self) -> Result<Self::Result> {
+        let mut result = self.inner.execute().await?;
+        loop {
+            let locks = result.take_locks();
+            if locks.is_empty() {
+                return Ok(result);
+            }
+
+            if self.backoff.is_none() {
+                return Err(Error::ResolveLockError);
+            }
+
+            let pd_client = self.pd_client.clone();
+            if resolve_locks(locks, pd_client.clone()).await? {
+                result = self.inner.execute().await?;
+            } else {
+                let mut clone = self.clone();
+                match clone.backoff.next_delay_duration() {
```

Comment:
The `clone` is always cloned from `self`, thus the status of `backoff` is always at its initial state.